soc/xtensa/intel_adsp/ace_v1x/power.c

```diff
@@ -26,17 +40,97 @@
  * (each bit corresponds to one ebb)
  * @param response_to_ipc       flag if ipc response should be send during power down
  */
-extern void ace_power_down(bool disable_lpsram, uint32_t *hpsram_pg_mask,
+extern void power_down(bool disable_lpsram, uint32_t *hpsram_pg_mask,
 			   bool response_to_ipc);
 
+/* NOTE: This struct will grow with all values that have to be stored for
+ * proper cpu restore after PG.
+ */
+struct core_state {
+	uint32_t a0;
+	uint32_t a1;
+	uint32_t vecbase;
+	uint32_t excsave2;
+	uint32_t excsave3;
+	uint32_t thread_ptr;
+	uint32_t intenable;
+};
+
+static struct core_state core_desc[CONFIG_MP_NUM_CPUS] = { 0 };
+static struct core_state *desc_alias = &core_desc;
+
+struct lpsram_header {
+	uint32_t alt_reset_vector;
+	uint32_t adsp_lpsram_magic;
+	void *lp_restore_vector;
+	uint32_t reserved;
+	uint32_t slave_core_vector;
+	uint8_t rom_bypass_vectors_reserved[0xC00 - 0x14];
+};
+
+void dsp_pg_restore(void);
+
+/* TODO: Fill this function with all necessary operations for preparing to power gating */
+void power_gate(uint32_t core_id)
+{
+	struct lpsram_header *lpsheader =
+		(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
+
+	xthal_window_spill();
+	core_desc[core_id].vecbase = RSR("VECBASE");
+	core_desc[core_id].excsave2 = RSR("EXCSAVE2");
+	core_desc[core_id].excsave3 = RSR("EXCSAVE3");
+	core_desc[core_id].thread_ptr = RUR("THREADPTR");
+	__asm__ volatile("mov %0, a0" : "=r"(core_desc[core_id].a0));
+	__asm__ volatile("mov %0, a1" : "=r"(core_desc[core_id].a1));
+	lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
+	lpsheader->lp_restore_vector = (uint32_t) &dsp_pg_restore;
+	soc_cpus_active[core_id] = false;
+	xthal_dcache_all_writeback();
+	z_xt_ints_on(ALL_USED_INT_LEVELS_MASK);
+	k_cpu_idle();
+	__asm__ volatile("dsp_pg_restore:\n\t"
+	"  movi  a0, 0\n\t"
+	"  movi  a1, 1\n\t"
+	"  wsr   a1, WINDOWSTART\n\t"
+	"  wsr   a0, WINDOWBASE\n\t"
+	"  rsync\n\t");
+
+	{
```

Comment:
Why this compound statement ?

Reply:
The compiler sees _power_gate()_ as one, but it works like two separate functions. At this point we have never return from _k_cpu_idle()_. We get here from ROM because of this:

```c
lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
lpsheader->lp_restore_vector = (uint32_t) &dsp_pg_restore;
```

With enabled power gating, core during waiti goes down. Memory has not been changed but all processor registers have been reset. In this way, I force all variables used after the reset to have a smaller (local) scope.

Reply:
Not following it, why do you need to force this variables to have a smaller scope ? The function returns at the end of this block, I think I am missing something.